### ISSUE TITLE
timeout=False puts the socket in non-blocking mode, causing weird errors.

### ISSUE THREAD

**Issue by offbyone** [kennethreitz/requests#2699]
I'm receiving a strange connection error using Requests 2.7.0
connecting to an HTTP service as part of an integration test. Below is
the call stack when I call into `session.request` in the API (for
reasons of NDA I can't provide the rest of the callstack).

The error is below in full detail, but in short, send() in requests is emitting `ConnectionError: ('Connection aborted.', error(115, 'Operation now in progress'))`

Python=2.7
Requests=2.7.0

The parameters being passed in to this are:

verb=POST
url=(an HTTP url for an internal service)
data=(a dict)
verify=False
cert=None
headers={'Content-Encoding': 'amz-1.0', 'Connection': 'keep-alive', 'Accept': 'application/json, text/javascript, _/_', 'User-Agent': 'A thing', 'Host': 'hostname', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache', 'Content-Type': 'application/json'}   # plus some service headers used by our service
timeout=False
proxies=Noneﬁ

> package-cache/packages/Requests/lib/python2.7/site-packages/requests/api.py:50: in request
>     response = session.request(method=method, url=url, *_kwargs)
> package-cache/packages/Requests/lib/python2.7/site-packages/requests/sessions.py:465: in request
>     resp = self.send(prep, *_send_kwargs)
> package-cache/packages/Requests/lib/python2.7/site-packages/requests/sessions.py:573: in send
>     r = adapter.send(request, **kwargs)
> ---
> 
> self = <requests.adapters.HTTPAdapter object at 0x7ff845dccd50>, request = <PreparedRequest [POST]>, stream = False, timeout = <requests.packages.urllib3.util.timeout.Timeout object at 0x7ff845dd0cd0>, verify = False, cert = None
> proxies = {}
> 
> ```
> def send(self, request, stream=False, timeout=None, verify=True, cert=None, proxies=None):
>     """Sends PreparedRequest object. Returns Response object.
> 
>         :param request: The :class:`PreparedRequest <PreparedRequest>` being sent.
>         :param stream: (optional) Whether to stream the request content.
>         :param timeout: (optional) How long to wait for the server to send
>             data before giving up, as a float, or a (`connect timeout, read
>             timeout <user/advanced.html#timeouts>`_) tuple.
>         :type timeout: float or tuple
>         :param verify: (optional) Whether to verify SSL certificates.
>         :param cert: (optional) Any user-provided SSL certificate to be trusted.
>         :param proxies: (optional) The proxies dictionary to apply to the request.
>         """
> 
>     conn = self.get_connection(request.url, proxies)
> 
>     self.cert_verify(conn, request.url, verify, cert)
>     url = self.request_url(request, proxies)
>     self.add_headers(request)
> 
>     chunked = not (request.body is None or 'Content-Length' in request.headers)
> 
>     if isinstance(timeout, tuple):
>         try:
>             connect, read = timeout
>             timeout = TimeoutSauce(connect=connect, read=read)
>         except ValueError as e:
>             # this may raise a string formatting error.
>             err = ("Invalid timeout {0}. Pass a (connect, read) "
>                    "timeout tuple, or a single float to set "
>                    "both timeouts to the same value".format(timeout))
>             raise ValueError(err)
>     else:
>         timeout = TimeoutSauce(connect=timeout, read=timeout)
> 
>     try:
>         if not chunked:
>             resp = conn.urlopen(
>                 method=request.method,
>                 url=url,
>                 body=request.body,
>                 headers=request.headers,
>                 redirect=False,
>                 assert_same_host=False,
>                 preload_content=False,
>                 decode_content=False,
>                 retries=self.max_retries,
>                 timeout=timeout
>             )
>         # Send the request.
>         else:
>             if hasattr(conn, 'proxy_pool'):
>                 conn = conn.proxy_pool
> 
>             low_conn = conn._get_conn(timeout=timeout)
> 
>             try:
>                 low_conn.putrequest(request.method,
>                                     url,
>                                     skip_accept_encoding=True)
> 
>                 for header, value in request.headers.items():
>                     low_conn.putheader(header, value)
> 
>                 low_conn.endheaders()
> 
>                 for i in request.body:
>                     low_conn.send(hex(len(i))[2:].encode('utf-8'))
>                     low_conn.send(b'\r\n')
>                     low_conn.send(i)
>                     low_conn.send(b'\r\n')
>                 low_conn.send(b'0\r\n\r\n')
> 
>                 r = low_conn.getresponse()
>                 resp = HTTPResponse.from_httplib(
>                     r,
>                     pool=conn,
>                     connection=low_conn,
>                     preload_content=False,
>                     decode_content=False
>                 )
>             except:
>                 # If we hit any problems here, clean up the connection.
>                 # Then, reraise so that we can handle the actual exception.
>                 low_conn.close()
>                 raise
>             else:
>                 # All is well, return the connection to the pool.
>                 conn._put_conn(low_conn)
> 
>     except (ProtocolError, socket.error) as err:
> ```
> 
> > ```
> >       raise ConnectionError(err, request=request)
> > ```
> > 
> > E           ConnectionError: ('Connection aborted.', error(115, 'Operation now in progress'))
> 
> cert       = None
> chunked    = False
> conn       = <requests.packages.urllib3.connectionpool.HTTPConnectionPool object at 0x7ff844c68c10>
> err        = ProtocolError('Connection aborted.', error(115, 'Operation now in progress'))
> proxies    = {}
> request    = <PreparedRequest [POST]>
> self       = <requests.adapters.HTTPAdapter object at 0x7ff845dccd50>
> stream     = False
> timeout    = <requests.packages.urllib3.util.timeout.Timeout object at 0x7ff845dd0cd0>
> url        = '/'
> verify     = False
> 
> package-cache/packages/Requests/lib/python2.7/site-packages/requests/adapters.py:415: ConnectionError

**Comment by Lukasa** [comment]
(Hi Chris!)

Well this is weird. Are you passing any socket options?

**Comment by offbyone** [comment]
None explicitly, no, and I checked in pdb; the options are the `default_socket_options`

**Comment by offbyone** [comment]
Basically, I've bottomed out on my socket-related knowhow. This code works for other services, so it's certainly possible that it's something related to this particular service endpoint, but I don't know what it would be.

**Comment by Lukasa** [comment]
Ok.

So, my basic understanding of this error is as follows. First, the error in question is EINPROGRESS (errno 115). I've seen this happen with a socket in non-blocking mode on a `connect()` call: it's very common in that case. However, if you're not passing explicit socket options then we aren't putting the socket in non-blocking mode, and it's hard to see why you'd do that (lots of stuff would go wrong if you did).

Can we clarify some stuff? Are you running in a deployment with PyOpenSSL, or just with the standard ssl library?

**Comment by offbyone** [comment]
I believe without PyOpenSSL, but... checking now...

As far as I know the only instance of PyOpenSSL on my sys.path is in requests/contrib, which I assume is not what you're talking about.

**Comment by Lukasa** [comment]
> As far as I know the only instance of PyOpenSSL on my sys.path is in requests/contrib, which I assume is not what you're talking about.

Correct. =) What I want to know is whether PyOpenSSL the package is accessible: that is, that running `python -c 'import OpenSSL'` fails in some way.

So, in that case I recommend running `strace` on your process. If the bug is readily reproducible then this will be particularly useful, because we can see the failing syscall, but I also want to see all syscalls relating to the socket in question. Let's see if it's actually being set to non-blocking mode or not.

**Comment by Lukasa** [comment]
(To be clear, setting a socket nonblocking will involve a `fnctl` call that can be seen in strace.)

**Comment by offbyone** [comment]
Okay, gimme a few; it's a 200Mb strace file all-in; narrowing this down will take me a bit.

**Comment by Lukasa** [comment]
No rush. =)

**Comment by offbyone** [comment]
```
socket(PF_NETLINK, SOCK_RAW, 0)         = 10
bind(10, {sa_family=AF_NETLINK, pid=0, groups=00000000}, 12) = 0
getsockname(10, {sa_family=AF_NETLINK, pid=13834, groups=00000000}, [12]) = 0
sendto(10, "\24\0\0\0\26\0\1\3s\217\272U\0\0\0\0\0\0\0\0", 20, 0, {sa_family=AF_NETLINK, pid=0, groups=00000000}, 12) = 20
recvmsg(10, {msg_name(12)={sa_family=AF_NETLINK, pid=0, groups=00000000}, msg_iov(1)=[{"D\0\0\0\24\0\2\0s\217\272U\n6\0\0\2\10\200\376\1\0\0\0\10\0\1\0\177\0\0\1"..., 4096}], msg_controllen=0, msg_flags=0}, 0) = 220
recvmsg(10, {msg_name(12)={sa_family=AF_NETLINK, pid=0, groups=00000000}, msg_iov(1)=[{"@\0\0\0\24\0\2\0s\217\272U\n6\0\0\n\200\200\376\1\0\0\0\24\0\1\0\0\0\0\0"..., 4096}], msg_controllen=0, msg_flags=0}, 0) = 192
recvmsg(10, {msg_name(12)={sa_family=AF_NETLINK, pid=0, groups=00000000}, msg_iov(1)=[{"\24\0\0\0\3\0\2\0s\217\272U\n6\0\0\0\0\0\0\1\0\0\0\24\0\1\0\0\0\0\0"..., 4096}], msg_controllen=0, msg_flags=0}, 0) = 20
close(10)                               = 0
socket(PF_FILE, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 10
connect(10, {sa_family=AF_FILE, path="/var/run/nscd/socket"}, 110) = 0
sendto(10, "\2\0\0\0\r\0\0\0\6\0\0\0hosts\0", 18, MSG_NOSIGNAL, NULL, 0) = 18
poll([{fd=10, events=POLLIN|POLLERR|POLLHUP}], 1, 5000) = 1 ([{fd=10, revents=POLLIN|POLLHUP}])
recvmsg(10, {msg_name(0)=NULL, msg_iov(2)=[{"hosts\0", 6}, {"H\344\17\0\0\0\0\0", 8}], msg_controllen=24, {cmsg_len=20, cmsg_level=SOL_SOCKET, cmsg_type=SCM_RIGHTS, {11}}, msg_flags=MSG_CMSG_CLOEXEC}, MSG_CMSG_CLOEXEC) = 14
mmap(NULL, 1041480, PROT_READ, MAP_SHARED, 11, 0) = 0x7f866d0b6000
close(11)                               = 0
close(10)                               = 0
socket(PF_INET, SOCK_STREAM, IPPROTO_TCP) = 10
setsockopt(10, SOL_TCP, TCP_NODELAY, [1], 4) = 0
fcntl(10, F_GETFL)                      = 0x2 (flags O_RDWR)
fcntl(10, F_SETFL, O_RDWR|O_NONBLOCK)   = 0
connect(10, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("10.12.164.47")}, 16) = -1 EINPROGRESS (Operation now in progress)
close(10)                               = 0
```

So, it looks like something _is_ setting it.

**Comment by Lukasa** [comment]
Fun!

So, the `setsockopt` call is us setting the "default socket options". I don't believe we call `setblocking` in our code. Out of interest, does any of yours make that call?

One way or another, it's happening somewhere between the point where we create the socket and call connect. In a vanilla requests/urllib3 that's really only a very small number of lines of code: requests.packages.urllib3.util.connection:68 to requests.packages.urllib3.util.connection:78.

One thing that might be worth doing is running in pdb and setting a breakpoint there, and also open strace in another window for that program, then step through and see if you can see where the fcntl call is coming from.

**Comment by offbyone** [comment]
I'll do that & report back. For what it's worth, this is the code calling requests:

```
    reply = requests.request(verb.lower(), uri.geturl(), data=http_data,
                             verify=verify_cert, cert=cert, headers=headers,
                             auth=http_auth, timeout=self._timeout,
                             proxies=self._proxies)
```

Doesn't seem to be anything there about sockets :/

**Comment by Lukasa** [comment]
No, I doubt there is, but I worry about monkeypatching and other kinds of fun. =D No rush, take your time on this.

**Comment by offbyone** [comment]
Based on the strace, it looks like socket.settimeout is doing it:

```
write(1, "-> sock.settimeout(timeout)\n", 28) = 28
ioctl(0, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(1, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(1, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
rt_sigprocmask(SIG_BLOCK, [INT], [], 8) = 0
ioctl(0, TIOCGWINSZ, {ws_row=0, ws_col=0, ws_xpixel=0, ws_ypixel=0}) = 0
ioctl(0, TIOCSWINSZ, {ws_row=0, ws_col=0, ws_xpixel=0, ws_ypixel=0}) = 0
ioctl(0, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig -icanon -echo ...}) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigaction(SIGWINCH, {0x7fafa61156a0, [], SA_RESTORER|SA_RESTART, 0x7fafb36daae0}, {SIG_DFL, [], SA_RESTORER, 0x7fafb36daae0}, 8) = 0
rt_sigprocmask(SIG_BLOCK, [INT], [], 8) = 0
write(1, "(Pdb) ", 6)                   = 6
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
select(1, [0], NULL, NULL, NULL)        = 1 (in [0])
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
read(0, "\n", 1)                        = 1
write(1, "\n", 1)                       = 1
```

**Comment by Lukasa** [comment]
That strace doesn't appear to have any fcntl calls in it at all...

**Comment by offbyone** [comment]
Shit. I missed them on paste. One sec.

**Comment by offbyone** [comment]
```
write(1, "-> sock.settimeout(timeout)\n", 28) = 28
ioctl(0, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(1, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(1, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
rt_sigprocmask(SIG_BLOCK, [INT], [], 8) = 0
ioctl(0, TIOCGWINSZ, {ws_row=0, ws_col=0, ws_xpixel=0, ws_ypixel=0}) = 0
ioctl(0, TIOCSWINSZ, {ws_row=0, ws_col=0, ws_xpixel=0, ws_ypixel=0}) = 0
ioctl(0, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig -icanon -echo ...}) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigaction(SIGWINCH, {0x7fafa61156a0, [], SA_RESTORER|SA_RESTART, 0x7fafb36daae0}, {SIG_DFL, [], SA_RESTORER, 0x7fafb36daae0}, 8) = 0
rt_sigprocmask(SIG_BLOCK, [INT], [], 8) = 0
write(1, "(Pdb) ", 6)                   = 6
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
select(1, [0], NULL, NULL, NULL)        = 1 (in [0])
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
read(0, "\n", 1)                        = 1
write(1, "\n", 1)                       = 1
rt_sigprocmask(SIG_BLOCK, [INT], [], 8) = 0
ioctl(0, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig icanon echo ...}) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigaction(SIGWINCH, {SIG_DFL, [], SA_RESTORER, 0x7fafb36daae0}, {0x7fafa61156a0, [], SA_RESTORER|SA_RESTART, 0x7fafb36daae0}, 8) = 0
fcntl(10, F_GETFL)                      = 0x2 (flags O_RDWR)
fcntl(10, F_SETFL, O_RDWR|O_NONBLOCK)   = 0
write(1, "> /home/chrisros/workspace/CDSer"..., 199) = 199
write(1, "-> if source_address:\n", 22) = 22
```

**Comment by offbyone** [comment]
Note the prompt boundaries for the pdb prompt; that's how I'm triangulating

**Comment by Lukasa** [comment]
_Aha_! Check this weaselly section of the Python docs (emphasis mine):

> Set a timeout on blocking socket operations. The value argument can be a nonnegative float expressing seconds, or None. If a float is given, subsequent socket operations will raise a timeout exception if the timeout period value has elapsed before the operation has completed. Setting a timeout of None disables timeouts on socket operations. **s.settimeout(0.0) is equivalent to s.setblocking(0)**; s.settimeout(None) is equivalent to s.setblocking(1).

So, what's the value of `self._timeout` in this call?

**Comment by sigmavirus24** [comment]
:clap: @Lukasa

**Comment by offbyone** [comment]
`False`

**Comment by Lukasa** [comment]
Yeah, that's not a valid value for that argument. ;) _And_, when evaluated in a numerical context, is equivalent to `0`, which is _probably_ the source of this bug.

Want to try catching that at your level and seeing if the problem goes away? If it does, we should consider whether we can police this in the API.

**Comment by Lukasa** [comment]
Heh, yup, the `Timeout` object coerces its argument to float, and in Python:

``` python
>>> float(False)
0.0
```

Sometimes I hate this language.

**Comment by sigmavirus24** [comment]
> Want to try catching that at your level and seeing if the problem goes away? If it does, we should consider whether we can police this in the API.

That would be a 3.0.0 change (to raise a ValueError if it isn't `None`, `tuple`, `int`, or `float`).

**Comment by Lukasa** [comment]
@sigmavirus24 Sounds good to me.

Separately, urllib3 may want a fix for this, because it is _also_ affected. /cc @shazow

**Comment by shazow** [comment]
What are we fixing? in urllib3, `timeout=False` has semantic meaning iirc.

**Comment by Lukasa** [comment]
@sigmavirus24 Actually, now that I think about it that's overly aggressive. Generally, things that are `float`-y or `int`-y should be totally acceptable. I think we just explicitly want to forbid `True`/`False`, because they happen to be `int`-y but really shouldn't be.

**Comment by Lukasa** [comment]
@shazow Not according to the docstring for the `Timeout` class, which is where this happens:

```
    Timeouts can be defined as a default for a pool::
        timeout = Timeout(connect=2.0, read=7.0)
        http = PoolManager(timeout=timeout)
        response = http.request('GET', 'http://example.com/')
    Or per-request (which overrides the default for the pool)::
        response = http.request('GET', 'http://example.com/', timeout=Timeout(10))
    Timeouts can be disabled by setting all the parameters to ``None``::
        no_timeout = Timeout(connect=None, read=None)
        response = http.request('GET', 'http://example.com/, timeout=no_timeout)
    :param total:
        This combines the connect and read timeouts into one; the read timeout
        will be set to the time leftover from the connect attempt. In the
        event that both a connect timeout and a total are specified, or a read
        timeout and a total are specified, the shorter timeout will be applied.
        Defaults to None.
    :type total: integer, float, or None
    :param connect:
        The maximum amount of time to wait for a connection attempt to a server
        to succeed. Omitting the parameter will default the connect timeout to
        the system default, probably `the global default timeout in socket.py
        <http://hg.python.org/cpython/file/603b4d593758/Lib/socket.py#l535>`_.
        None will set an infinite timeout for connection attempts.
    :type connect: integer, float, or None
    :param read:
        The maximum amount of time to wait between consecutive
        read operations for a response from the server. Omitting
        the parameter will default the read timeout to the system
        default, probably `the global default timeout in socket.py
        <http://hg.python.org/cpython/file/603b4d593758/Lib/socket.py#l535>`_.
        None will set an infinite timeout.
    :type read: integer, float, or None
```

Now, admittedly, the docstring does not allow `True`/`False`, but those can be integer or float if evaluated in that context, so you may want to police them. I can kinda see that `timeout=False` putting the socket in non-blocking mode might make sense in some kind of weird world, but I cannot see how `timeout=True` should put the socket into one-second-timeout mode.

**Comment by offbyone** [comment]
`timeout=None` works for the 'set nonblocking' concept better than `timeout=False` imo

**Comment by shazow** [comment]
Erm I'm thinking of retries=False.

**Comment by Lukasa** [comment]
@offbyone You'd have thought so, right? Except again, the Python `socket` library is all stupid:

>  s.settimeout(`None`) is equivalent to s.setblocking(1)

So...yeah.

**Comment by offbyone** [comment]
@Lukasa, this is working now that I'm policing the timeout on my side. For my end, this is fixed. Feel free to leave this open if you want to track the core issue, or close & make a separate one.

Thanks!

**Comment by Lukasa** [comment]
@offbyone Happy to help: this may be my new favourite bug report on this library. =D

I think we should do some policing of this on our end, so I'm keeping it open but renaming the issue.

**Comment by sigmavirus24** [comment]
@Lukasa I was thinking along the lines of

``` py
if not isinstance(timeout, (tuple, int, float)):
    raise ValueError()
```

If people have something that should behave like an int/float/tuple they can cast it to that before giving it to us IMO. That said, if we want to eventually allow `Timeout` objects then that might be problematic, but we've never had a desire to allow that so I don't see that being a problem.

**Comment by Lukasa** [comment]
@sigmavirus24 Ah, crap, that doesn't work. Observe the majesty of Python:

``` python
>>> isinstance(False, int)
True
```

I wonder if we should explicitly disallow anything that compares numerically equal to zero, because setting the socket to non-blocking mode is a wacky and bizarre thing to want to do. Alternatively, we could coerce anything that compares numerically equal to zero to `None` instead, which makes the concrete assertion that when you said `timeout=0` or `timeout=False` what you meant was "don't time out".

@shazow @kevinburke It's my assertion that we shouldn't let anyone set `0` as their timeout for either connection or read timeouts, because it doesn't mean what they think it means. We should either coerce to `None` or reject outright. Do either of you disagree?

**Comment by kevinburke** [comment]
Seems fine to me.

**Comment by offbyone** [comment]
If I understand this correctly, there are two not-quite-orthogonal concepts in play here. There's "is the socket blocking or non-blocking" and "if non-blocking, what is the timeout on the socket". Unless I'm mistaken, there's a design level bug where these two things are conflated for the convenience of a single argument.

(Note: am often mistaken)

_given_ that they're conflated inappropriately, if requests/urrllib works with both kinds of sockets then you should coerce and allow the value selection to toggle between these two concepts. If only blocking sockets are allowed then you should reject the values that would result in nonblocking (and vice versa).

**Comment by sigmavirus24** [comment]
@Lukasa we could also just explicitly check for boolean before hand so we know we've handled that case. Yes, it is ugly, but we're trying to hide complexity as it is (which is what @offbyone is explaining by thinking aloud through this problem) so we can probably continue to do that.

**Comment by Lukasa** [comment]
Only blocking sockets are allowed in requests: non-blocking sockets are explicitly not supported. That means @offbyone is suggesting we reject the value (`ValueError`, probably): does anyone think we should coerce instead?

**Comment by sigmavirus24** [comment]
ValueError is fine by me. I'm not in favor of silently doing things users might not expect. That leads to arguments and bug reports and pain.

**Comment by offbyone** [comment]
+1 and make sure the message is clear.

**Comment by shazow** [comment]
Yea, if `0` activates a mode that we don't yet support, then an informative ValueError makes sense.

**Comment by Lukasa** [comment]
@kevinburke There is one test (`test.with_dummyserver.test_connectionpool.TestConnectionPool.test_total_timeout`) that appears to expect a read timeout of 0 to work (specifically, to raise EAGAIN). This makes me think that maybe you wanted timeout of 0 to work for read. Is that right, or were you just asserting the actual behaviour of setting the timeout in that way?

**Comment by Lukasa** [comment]
Ping again @kevinburke.

**Comment by koliyo** [comment]
Same problem here, timeout=0 gives

```
.('Connection aborted.', error(36, 'Operation now in progress'))
```

really confusing error message, it should at least be documented in http://docs.python-requests.org/en/master/user/quickstart/#timeouts if zero timeout is not allowed

I do however think that zero timeout should be allowed

**Comment by Lukasa** [comment]
@kolis What would a zero timeout mean, semantically speaking?

**Comment by koliyo** [comment]
Sorry, ignore that comment, I wasn't thinking it through properly :)

**Comment by nateprewitt** [comment]
@Lukasa @sigmavirus24 Is this something we still want patched? [Here](https://github.com/nateprewitt/requests/commit/1b5b8e4cd8344ed6031fd758ee802e2dfce81e9e) is a quick pass at it, but I'm not sure if this should be PR'd against master or 3.0. There's also an minor side issue with urllib3 that it raises a `ValueError` for values < 0 stating that timeout can be set to 0 or greater.

**Comment by nateprewitt** [comment]
@Lukasa @sigmavirus24 Is this something we still want patched? [Here](https://github.com/nateprewitt/requests/commit/6b7dd6ae4265cc308e6a7cdbc17fe4812e6caa67) is a quick pass at it, but I'm not sure if this should be PR'd against master or 3.0. There's also a minor side issue with urllib3 that it raises a `ValueError` for values < 0 stating that timeout can be set to 0 or greater.

**Comment by Lukasa** [comment]
I think it'd be better to fix this in urllib3 rather than in requests.

**Comment by nateprewitt** [comment]
Ok, that sounds good to me, I was raising this here based on [this comment](https://github.com/kennethreitz/requests/issues/2699#issuecomment-126498936). Would you prefer I open a ticket against urllib3 with a PR, or reference this ticket?

**Comment by Lukasa** [comment]
You can just to straight to urllib3 with a PR, no need for a new ticket.

**Comment by nateprewitt** [comment]
Ok will do, and to make sure before I open this up, urllib3 is not supporting non-blocking requests, correct? I'm not clear based on the [previous comments](https://github.com/kennethreitz/requests/issues/2699#issuecomment-136610616) if this was a requests specific decision or applies to urllib3 as well.

Edit: Actually according to [this line](https://github.com/shazow/urllib3/blob/master/test/test_util.py#L307) in urllib3 from 2013, True and False are explicitly defined to represent 1.0 and 0.0 respectively.

**Comment by Lukasa** [comment]
Yup, but I think that test is wrong. Kevin wrote the test and IIRC he agrees with me. I feel OK about tightening this up. Nonblocking makes no sense here: urllib3 is just not equipped to deal with it.

**Comment by nateprewitt** [comment]
Ok great! I'll open a PR shortly.

**Comment by sigmavirus24** [comment]
Since this merged in urllib3, I think we will end up benefiting from this in the next 2.X.0 release. Any objections to closing this?

**Comment by nateprewitt** [comment]
This should be addressed now in 2.12.0.
